presidio-recognizers-store/cmd/presidio-recognizers-store/main.go

```diff
@@ -0,0 +1,348 @@
+	log.CreateLogger(settings.LogLevel)
+
+	svc := services.New(settings)
+
+	if settings.RedisURL != "" {
+		recognizersStore = svc.SetupCache()
+	}
+
+	lis, s := rpc.SetupClient(settings.GrpcPort)
+
+	types.RegisterRecognizersStoreServiceServer(s, &server{})
+	reflection.Register(s)
+
+	if err := s.Serve(lis); err != nil {
+		log.Fatal(err.Error())
+	}
+}
+
+// setValue sets the recognizers array in the store and updates the last update
+// timestamp
+func setValue(value string, timestamp string) error {
+	err := recognizersStore.Set(recognizersKey, value)
+	if err != nil {
+		log.Error(err.Error())
+		return err
+	}
+
+	log.Info("Updating timestamp: " + timestamp)
+	err = recognizersStore.Set(timestampKey, timestamp)
+	if err != nil {
+		log.Error(err.Error())
+		return err
+	}
+
+	return nil
+}
+
+func getExistingRecognizers() ([]types.PatternRecognizer, error) {
+	recognizersArr := make([]types.PatternRecognizer, 0)
+	existingItems, err := recognizersStore.Get(recognizersKey)
+	if err == nil && existingItems != "" {
+		log.Info("Found existing recognizers...")
+		var recognizers []types.PatternRecognizer
+		err = json.Unmarshal([]byte(existingItems), &recognizers)
+		if err != nil {
+			return nil, err
+		}
+		recognizersArr = append(recognizersArr, recognizers...)
+	}
+	return recognizersArr, nil
+}
+
+// InsertOrUpdateRecognizer inserts (or updated) a recognizer to the store
+func insertOrUpdateRecognizer(value string, isUpdate bool) error {
+	if recognizersStore == nil {
+		return errors.New("cache is missing")
+	}
+
+	existingRecognizersArr, err := getExistingRecognizers()
+	if err != nil {
+		return err
+	}
+
+	var newRecognizer types.PatternRecognizer
+	json.Unmarshal([]byte(value), &newRecognizer)
+
+	if isUpdate == false {
+		// Insert new item, verify doesn't exists and append
+		for _, element := range existingRecognizersArr {
+			if element.Name == newRecognizer.Name {
+				return errors.New(
+					"custom pattern recognizer with name '" +
+						element.Name + "' already exists")
+			}
+		}
+
+		// verified, append to result array
+		existingRecognizersArr = append(existingRecognizersArr, newRecognizer)
+	} else {
+		// Update, find and update
+		found := false
+		for i, element := range existingRecognizersArr {
+			if element.Name == newRecognizer.Name {
+				found = true
+
+				log.Info("Found recognizer to be updated")
+				// remove the 'old' value
+				existingRecognizersArr = append(existingRecognizersArr[:i],
+					existingRecognizersArr[i+1:]...)
+				// append the new one
+				existingRecognizersArr = append(existingRecognizersArr,
+					newRecognizer)
+				break
+			}
+		}
+		if found == false {
+			return errors.New("custom pattern recognizer with name '" +
+				newRecognizer.Name + "' was not found")
+		}
+	}
+
+	recognizersBytes, _ := json.Marshal(existingRecognizersArr)
+	currentTimeStr := strconv.FormatInt(time.Now().Unix(), 10)
+
+	return setValue(string(recognizersBytes), currentTimeStr)
```

Comment:
An option to resolve this would be to move to using the [Redis list](https://redis.io/topics/data-types#lists) datatype

Reply:
Or look at using a more transnational store like PostgresSQL, K8s CRDs etc

Reply:
I've elaborated why I think Redis is a good option [here](https://github.com/Microsoft/presidio/pull/104#discussion_r268660677).
Since in this implementation I chose to keep all the recognizers under a single key 'custom_recognizers' it makes it very easy to retrieve with a single query (although insert/update/delete are harder) and it helps with the timestamp creation (which I will change to etag as discussed above). 

Moving to Redis list DS will not completely remove the possibility of race, although it will decrease the probability.
Another assumption we can leverage, is that writing recognizers is _very very rare_ and we can compromise on the time it takes for a write operation to finish.
Hence, I suggest adding a mutex in insert/update/delete operations, this will completely remove the race issue and at the same time wont affect the 'get' performance we strive to keep as fast as possible.

Reply:
Mutex would work as long as there is only ever one instance of the store service running. 

I think either using `Transactions` or `List` operations (Less sure on this one) in redis would allow you to avoid the race condition while sticking with this approach ~~BUT I've never tried these out.~~

https://redis.io/topics/transactions
https://stackoverflow.com/questions/40999446/how-to-create-redis-transaction-in-go-using-go-redis-redis-package

I tried this out and it works nicely and uses the transaction removes the race for multi deployments. 

In the code below the first go routine writes every 1 second to the key and the second reads then pauses for 5 seconds before attempting a while (simulating the read then write in your code). The transaction fails when twice as the key is changes between the read and write then succeeds on the last try as go routine 1 has stopped updating the key. 

```
15:48 $ go build . && ./go-redis-list 
#PONG <nil>
2019/03/26 15:49:10 #1ROUTINE: Key set outside transation to 2
2019/03/26 15:49:11 #1ROUTINE: Key set outside transation to 3
2019/03/26 15:49:12 #1ROUTINE: Key set outside transation to 4
2019/03/26 15:49:13 #1ROUTINE: Key set outside transation to 5
2019/03/26 15:49:14 #1ROUTINE: Key set outside transation to 6
2019/03/26 15:49:15 #2ROUTINE: Failed to write redis: transaction failed 
2019/03/26 15:49:15 #1ROUTINE: Key set outside transation to 7
2019/03/26 15:49:16 #1ROUTINE: Key set outside transation to 8
2019/03/26 15:49:17 #1ROUTINE: Key set outside transation to 9
2019/03/26 15:49:18 #1ROUTINE: Key set outside transation to 10
2019/03/26 15:49:20 #2ROUTINE: Failed to write redis: transaction failed 
2019/03/26 15:49:25 #2ROUTINE: Transaction updated key to: get RECOGKEY: get RECOGKEY: 10-after

```

```
package main

import (
	"fmt"
	"log"
	"sync"
	"time"

	"github.com/go-redis/redis"
)

func main() {
	client := redis.NewClient(&redis.Options{
		Addr:     "localhost:6379",
		Password: "", // no password set
		DB:       0,  // use default DB
	})

	pong, err := client.Ping().Result()
	fmt.Println(pong, err)
	// Output: PONG <nil>

	key := "RECOGKEY"

	client.Set(key, "1", time.Duration(time.Second*360))

	wg := sync.WaitGroup{}
	wg.Add(2)

	go func() {
		i := 1
		for {
			i++

			if i > 10 {
				break
			}

			client.Set(key, i, time.Duration(time.Second*360))
			log.Printf("#1ROUTINE: Key set outside transation to %v", i)

			time.Sleep(1 * time.Second)
		}
		wg.Done()
	}()

	go func() {
		for {
			err := client.Watch(func(tx *redis.Tx) error {
				n := tx.Get(key).String()

				_, err = tx.Pipelined(func(pipe redis.Pipeliner) error {
					time.Sleep(time.Second * 5)
					pipe.Set(key, n+"-after", 0)
					return nil
				})
				return err
			}, key)

			if err != nil {
				log.Printf("#2ROUTINE: Failed to write %+v \n", err)
				continue
			}

			value := client.Get(key)
			log.Printf("#2ROUTINE: Transaction updated key to: %+v", value)

			break
		}
		wg.Done()
	}()

	wg.Wait()
}
```

Reply:
(Not saying you have to do this just was interested to see if it was possible to do it with Redis as I'd not used their transations before)

Reply:
The caveat with Redis transactions is that you can't get a value, process it and then set it.
As the Get command will be 'Queued', meaning the value won't be available until Exec is called.

There probably other solutions (such as lua scripts).

Reply:
To sum this conversation:
We expect very few changes to the config and even less possibilities for such a race condition. 
The mutex is a simple yet effective solution for 1 replica, if someone needs to scale this microservice, it is still possible, the race can then still happen but with even lower possibility as this means two commands arrived at the same time to two different replicas.

Marking as resolved